src/EventLoop.cpp

```diff
@@ -358,44 +441,78 @@ auto requests_accept_async(
-        event_loop->m_grabbed_requests.swap(
-            event_loop->m_pending_requests);
+        event_loop->m_grabbed_requests.swap(event_loop->m_pending_requests);
     }
 
-    for (auto& request : event_loop->m_grabbed_requests) {
-        auto& raw_request_handle = request.m_request_handle;
-
-        auto curl_code = curl_multi_add_handle(event_loop->m_cmh, raw_request_handle->m_curl_handle);
-
+    for (auto& request_handle : event_loop->m_grabbed_requests)
+    {
+        auto& raw_request = *request_handle;
+        
+        // If there's a response wait time, we'll need to create a ResponseWaitTimeWrapper and add
+        // it to the wrapper multiset so it can be handled in the event it takes too long to respond.
+        if (const auto& response_wait_time_opt = raw_request.GetResponseWaitTime(); response_wait_time_opt.has_value())
+        {
+            // Get the current timepoint from libuv (it caches the current timepoint at the beginning
+            // of every cycle through the event loop, so this should not be expensive).
+            auto time = uv_now(event_loop->m_loop);
+            
+            const auto request_timeout = response_wait_time_opt.value();
+            auto next_timepoint = time + static_cast<uint64_t>(request_timeout.count());
+        
+            auto& wrappers = event_loop->m_response_wait_time_wrappers;
+            
+            if (wrappers.empty() || next_timepoint < wrappers.begin()->GetData().m_timeout_time)
+            {
+                uv_timer_stop(&event_loop->m_response_timer);
+    
+                uv_timer_start(
+                    &event_loop->m_response_timer,
+                    on_response_wait_time_expired_callback,
+                    static_cast<uint64_t>(request_timeout.count()),
+                    0);
+            }
+        
+            auto iterator = event_loop->m_response_wait_time_wrappers.emplace(next_timepoint, request_handle.createSharedRequestOnHeap());
+            request_handle->setTimeoutIterator(iterator);
+        }
+    
+        // Put the shared_ptr to the SharedRequest on the heap and set curl handle's data to point to the pointer,
+        // so its lifetime is maintained even after exiting this function
+        raw_request.setSharedPointerOnCurlHandle(request_handle.createSharedRequestOnHeap());
+        
+        auto curl_code = curl_multi_add_handle(event_loop->m_cmh, raw_request.m_curl_handle);
+    
         if(curl_code != CURLM_OK && curl_code != CURLM_CALL_MULTI_PERFORM)
         {
             /**
              * If curl_multi_add_handle fails then notify the user that the request failed to start
              * immediately.
              */
-            request->setCompletionStatus(CURLcode::CURLE_SEND_ERROR);
-            request->onComplete();
+            request_handle->setCompletionStatus(CURLcode::CURLE_SEND_ERROR);
+            request_handle->onComplete(*event_loop, *request_handle.createSharedRequestOnHeap());
```

Comment:
this looks like a memory leak to me, nothing will delete the shared_ptr on the heap, and thus not the shared request either

Reply:
Good point -- moving the shared pointer into onComplete so it doesn't leak